worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/segment/SegmentMapPartitionFileWriter.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.celeborn.service.deploy.worker.storage.DeviceMonitor;
+import org.apache.celeborn.service.deploy.worker.storage.MapPartitionDataWriter;
+import org.apache.celeborn.service.deploy.worker.storage.PartitionDataWriterContext;
+import org.apache.celeborn.service.deploy.worker.storage.StorageManager;
+
+/**
+ * Write the shuffle file in map partition format with segment granularity visibility. This means
+ * that the shuffle file should be written intermediate, allowing it to be read in segments rather
+ * than waiting for the entire shuffle file to be completed.
+ */
+public class SegmentMapPartitionFileWriter extends MapPartitionDataWriter {
+
+  public static final Logger logger = LoggerFactory.getLogger(SegmentMapPartitionFileWriter.class);
+
+  /**
+   * subPartitionId -> started (boolean). There are 3 cases: 1. If the subPartition key not exist,
+   * it indicates that the subPartition has not sent the {@link
+   * org.apache.celeborn.common.protocol.PbSegmentStart}. Therefore, shuffle data should not be
+   * written in this case. 2. If the subPartition key exists and the started value is true, it means
+   * that the subPartition has initiated the segment. In this situation, the next buffer for this
+   * subPartition will be the first buffer of the current segment. The information about the first
+   * buffer will be recorded in {@code MapFileMeta#subPartitionSegmentIndexes}. 3. If the
+   * subPartition key exists and the started value is false, it means that the subPartition has
+   * initiated the segment, but the next buffer for this subPartition is not the first buffer of the
+   * current segment.
+   */
+  private final Map<Integer, Boolean> subPartitionHasStartSegment;
+
+  // current buffer index per subPartition
+  private int[] subPartitionBufferIndex;
+
+  public SegmentMapPartitionFileWriter(
+      StorageManager storageManager,
+      AbstractSource workerSource,
+      CelebornConf conf,
+      DeviceMonitor deviceMonitor,
+      PartitionDataWriterContext writerContext)
+      throws IOException {
+    super(storageManager, workerSource, conf, deviceMonitor, writerContext);
+    this.subPartitionHasStartSegment = new HashMap<>();
+  }
+
+  @Override
+  public void pushDataHandShake(int numSubpartitions, int bufferSize) {
+    super.pushDataHandShake(numSubpartitions, bufferSize);
+    subPartitionBufferIndex = new int[numSubpartitions];
+    Arrays.fill(subPartitionBufferIndex, 0);
+    getFileMeta().setIsWriterClosed(false);
+    getFileMeta().setSegmentGranularityVisible(true);
+  }
+
+  @Override
+  public void write(ByteBuf data) throws IOException {
+    data.markReaderIndex();
+    int subPartitionId = data.readInt();
+    int attemptId = data.readInt();
+    int batchId = data.readInt();
+    int size = data.readInt();
+
+    if (!subPartitionHasStartSegment.containsKey(subPartitionId)) {
+      throw new IllegalStateException("This partition may not start a segment: " + subPartitionId);
```

Comment:
Could the exception message include `attemptId`, `batchId` and `size`? Otherwise this check could move before line 88.